src/main/java/fi/aalto/cs/apluscourses/ui/base/TreeView.java

```diff
@@ -0,0 +1,105 @@
+
+public class TreeView extends Tree {
+  private final transient Set<ActionListener> nodeAppliedListeners = ConcurrentHashMap.newKeySet();
+
+  @Nullable
+  protected transient volatile SelectableNodeViewModel<?> selectedItem = null; //NOSONAR
+
+  /**
+   * Construct an empty tree with no nodes and the root set to invisible.
+   */
+  public TreeView() {
+    setRootVisible(false);
+    getSelectionModel().setSelectionMode(TreeSelectionModel.SINGLE_TREE_SELECTION);
+    addTreeSelectionListener(new SelectionListener());
+    addMouseListener(new TreeMouseListener());
+  }
+
+  /**
+   * Set the model of the this tree to the given view model, or do nothing if the given view model
+   * is {@code null}.
+   */
+  public void setViewModel(@Nullable TreeViewModel viewModel) {
+    if (viewModel != null) {
+      setModel(new DefaultTreeModel(createNode(viewModel)));
+    }
+    selectedItem = null;
+  }
+
+  @NotNull
+  private static DefaultMutableTreeNode createNode(@NotNull TreeViewModel tree) {
+    List<? extends TreeViewModel> subtrees = tree.getSubtrees();
+    boolean allowsChildren = subtrees != null;
+    DefaultMutableTreeNode node = new DefaultMutableTreeNode(tree, allowsChildren);
+    if (allowsChildren) {
+      for (TreeViewModel subtree : subtrees) {
+        node.add(createNode(subtree));
+      }
+    }
+    return node;
+  }
+
+  public void addNodeAppliedListener(ActionListener listener) {
+    nodeAppliedListeners.add(listener);
+  }
+
+  public void removeNodeAppliedListener(ActionListener listener) {
+    nodeAppliedListeners.remove(listener);
+  }
+
+  protected class SelectionListener implements TreeSelectionListener {
+
+    @Override
+    public void valueChanged(@NotNull TreeSelectionEvent e) {
+      SelectableNodeViewModel<?> oldViewModel = getViewModelFromPath(e.getOldLeadSelectionPath());
+      if (oldViewModel != null) {
+        oldViewModel.setSelected(false);
+      }
+      SelectableNodeViewModel<?> newViewModel = getViewModelFromPath(e.getNewLeadSelectionPath());
+      if (newViewModel != null) {
+        newViewModel.setSelected(true);
+      }
+      selectedItem = newViewModel;
+    }
+
+    @Nullable
+    private SelectableNodeViewModel<?> getViewModelFromPath(@Nullable TreePath treePath) {
+      if (treePath == null || treePath.getPathCount() <= 1) {
+        return null; // Null for the root node, since it is hidden.
+      }
+      DefaultMutableTreeNode node = (DefaultMutableTreeNode) treePath.getLastPathComponent();
+      return (SelectableNodeViewModel<?>) node.getUserObject();
+    }
+  }
+
+  protected class TreeMouseListener extends MouseAdapter {
+
+    @Override
+    public void mouseClicked(@NotNull MouseEvent e) {
+      if (e.getClickCount() == 2 && selectedItem != null) {
+        ActionEvent actionEvent = new ActionEvent(this, ActionEvent.ACTION_PERFORMED, null);
+        nodeAppliedListeners.forEach(listener -> listener.actionPerformed(actionEvent));
+      }
+    }
+  }
```

Comment:
I forgot to add this myself, but an enter key listener (toggle expansion if non-leaf none, call node listeners if leaf node) would be a nice-to-have addition